### ISSUE TITLE
http: superfluous response.WriteHeader call

### ISSUE THREAD

**Issue by tsabi** [mattermost/mattermost#24991]
#### Summary
Many of our users are constantly in disconnected state.
This is not a security flaw (it has been requested in the comment to be stated).

#### Steps to reproduce
I am not sure.
Half of the users are fully online and everything works fine. And the other half of the users are in a disconnected state: they are logged in, but they have a blue bar on top indicating that they are in a disconnected state. Logged in users see disconnected users as logged in (solid green dot). Disconnected users see new messages and notifications when they refresh the website or chat app.

I couldn't figure out why one user was constantly disconnected and why others worked. Everyone uses it the same way from multiple devices, is in multiple rooms, etc. Some work and some don't. 

#### Observed behavior (that appears unintentional) 
I use nginx as a web proxy. On the mattermost server i use ubuntu 22.04 with latest updates, PostgreSQL 15.4-2.pgdg22.04+1, and Mattermost server 8.1.1-0.
I tried to delete the sessions from the system console, the same users were remained disconnected. I tried to do the same by cleaning the sessions table in SQL server.
I had seen that the number of allowed connections to the sql server was low, so I increased the number to 120, but this didn't solve this particular problem either.

I see this error message thousands of times on my server:
{
  "caller": "log/log.go:198",
  "level": "error",
  "msg": "http: superfluous response.WriteHeader call from github.com/mattermost/mattermost/server/v8/channels/web.(*responseWriterWrapper).WriteHeader (response_writer_wrapper.go:39)",
  "source": "httpserver",
  "timestamp": "2023-10-18 06:46:57.227 Z"
}

**Comment by amyblais** [comment]
Opened https://mattermost.atlassian.net/browse/MM-55012.

**Comment by agnivade** [comment]
Hello @tsabi - the `http: superfluous response.WriteHeader` error is harmless and is not related to your issue. This error happens when we call `w.WriteHeader` _after_ already writing the response with `w.Write`. It must be there in some parts of the codebase, which we would need to fix. 

I don't know why some users keep getting disconnected. I would suggest to look at your nginx config settings and if there is any timeout that can affect this. Or perhaps try to make a direct connection with Mattermost bypassing nginx and see if that fixes your issue.

**Comment by tsabi** [comment]
@agnivade 
I cannot see any problem in the nginx config, and direct connecting to the server is hard, because it is behind a firewall. I try to set up a VPN to direct connect, in the mean time could you please look at my nginx config? Thanks in advance!

```nginx
upstream proxy_mattermost {
   server 10.0.0.11:8065;
   keepalive 32;
}

server {
 listen 1.2.3.4:443 ssl http2;
 server_name chat.example.hu;
 charset utf-8;

 ssl_certificate /etc/letsencrypt/live/chat.example.hu/fullchain.pem;
 ssl_certificate_key /etc/letsencrypt/live/chat.example.hu/privkey.pem;
 ssl_trusted_certificate       /etc/letsencrypt/live/chat.example.hu/chain.pem;

 # Enable HTTP/2 Server Push
 http2_push_preload on;

 ssl_dhparam /etc/nginx/dhparams.pem;
 add_header        Alternate-Protocol  443:h2;

 # SSL
 ssl_session_timeout 1d;
 ssl_session_cache shared:SSL:20m;
 ssl_session_tickets off;

 # intermediate configuration
 ssl_protocols TLSv1.2 TLSv1.3;
 ssl_ciphers ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384;
 ssl_prefer_server_ciphers off;
 ssl_ecdh_curve secp384r1;

 # OCSP Stapling
 ssl_stapling on;
 ssl_stapling_verify on;
 resolver 10.0.0.10 1.1.1.1 1.0.0.1 8.8.8.8 8.8.4.4 208.67.222.222 208.67.220.220 valid=60s;
 resolver_timeout 2s;

 # security headers
 add_header X-Frame-Options "SAMEORIGIN" always;
 add_header X-XSS-Protection "1; mode=block" always;
 add_header X-Content-Type-Options "nosniff" always;
 add_header Strict-Transport-Security "max-age=63072000; includeSubDomains; preload" always;

 # Enable TLSv1.3's 0-RTT. Use $ssl_early_data when reverse proxying to
 # prevent replay attacks.
 #
 # @see: https://nginx.org/en/docs/http/ngx_http_ssl_module.html#ssl_early_data
 ssl_early_data on;
 add_header X-Early-Data $tls1_3_early_data;

 access_log /var/log/nginx/nginx_chat.example.hu.log main;
 error_log /var/log/nginx/nginx_error_chat.example.hu.log info;

 client_max_body_size 50M;
 client_body_buffer_size 16k;

   location ~ /api/v[0-9]+/(users/)?websocket$ {
       proxy_set_header Upgrade $http_upgrade;
       proxy_set_header Connection "upgrade";
       client_max_body_size 50M;
       proxy_set_header Host $http_host;
       proxy_set_header X-Real-IP $remote_addr;
       proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
       proxy_set_header X-Forwarded-Proto $scheme;
       proxy_set_header X-Frame-Options SAMEORIGIN;
       proxy_buffers 256 16k;
       proxy_buffer_size 16k;
       client_body_timeout 60;
       send_timeout 300;
       lingering_timeout 5;
       proxy_connect_timeout 90;
       proxy_send_timeout 300;
       proxy_read_timeout 90s;
       proxy_http_version 1.1;
       proxy_pass http://proxy_mattermost;
       proxy_ssl_server_name on;
   }

   location / {
       client_max_body_size 50M;
       proxy_set_header Connection "";
       proxy_set_header Host $http_host;
       proxy_set_header X-Real-IP $remote_addr;
       proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
       proxy_set_header X-Forwarded-Proto $scheme;
       proxy_set_header X-Frame-Options SAMEORIGIN;
       proxy_buffers 256 16k;
       proxy_buffer_size 16k;
       proxy_read_timeout 600s;
       proxy_cache mattermost_cache;
       proxy_cache_revalidate on;
       proxy_cache_min_uses 2;
       proxy_cache_use_stale timeout;
       proxy_cache_lock on;
       proxy_http_version 1.1;
       proxy_pass http://proxy_mattermost;
       proxy_ssl_server_name on;
   }
}
```

**Comment by agnivade** [comment]
Config looks okay to me. Only small thing is that the `proxy_ssl_server_name` directive is useless, unless you are connecting to an SSL backend which does not look to be in your case.

**Comment by tsabi** [comment]
Thanks @agnivade for your help!
Actually, I recently re-read the documentation and found a hint to set the CORS source (it was empty before). I set it and everything is fine since then. I mean, I think that solved it, because lately i just modified that setting.

**Comment by agnivade** [comment]
Was it the `AllowCorsFrom` setting? So if that's the case, it would have affected _all_ of your users not just some. That piece of information sent us down the wrong path :)

**Comment by tsabi** [comment]
I am also confused, I have no reasonable explanation. On my own machine if I logged into the web client first with my user it was fine, if I logged out and immediately logged in with my colleague's user it showed him offline. It was clearly a user dependent phenomenon but I couldn't figure out how.

I saw a small reference to this setting in the documentation, and after I changed the empty value to the web addresses I use, it was right away good for all users.
